devlib/target_runner.py

```diff
@@ -0,0 +1,267 @@
+
+    :param make_target: Lambda function for creating :class:`Target` based
+        object, defaults to :func:`lambda **kwargs: LinuxTarget(**kwargs)`.
+    :type make_target: func, optional
+
+    :Variable positional arguments: Forwarded to :class:`TargetRunner`.
+
+    :raises FileNotFoundError: if QEMU executable, kernel or initrd image cannot be found.
+    """
+
+    def __init__(self,
+                 qemu_settings,
+                 connection_settings=None,
+                 # pylint: disable=unnecessary-lambda
+                 make_target=lambda **kwargs: LinuxTarget(**kwargs),
+                 **args):
+        self.connection_settings = {
+            'host': '127.0.0.1',
+            'port': 8022,
+            'username': 'root',
+            'password': 'root',
+            'strict_host_check': False,
+        }
+
+        if connection_settings is not None:
+            self.connection_settings = self.connection_settings | connection_settings
+
+        qemu_args = {
+            'kernel_image': '',
+            'arch': 'aarch64',
+            'cpu_type': 'cortex-a72',
+            'initrd_image': '',
+            'mem_size': 512,
+            'num_cores': 2,
+            'num_threads': 2,
+            'cmdline': 'console=ttyAMA0',
+            'enable_kvm': True,
+        }
+
+        qemu_args = qemu_args | qemu_settings
+
+        qemu_executable = f'qemu-system-{qemu_args["arch"]}'
+        qemu_path = which(qemu_executable)
+        if qemu_path is None:
+            raise FileNotFoundError(f'Cannot find {qemu_executable} executable!')
+
+        if not os.path.exists(qemu_args["kernel_image"]):
+            raise FileNotFoundError(f'{qemu_args["kernel_image"]} does not exist!')
+
+        # pylint: disable=consider-using-f-string
+        qemu_cmd = '''\
+{} -kernel {} -append "{}" -m {} -smp cores={},threads={} -netdev user,id=net0,hostfwd=tcp::{}-:22 \
+-device virtio-net-pci,netdev=net0 --nographic\
+'''.format(
+            qemu_path,
+            qemu_args["kernel_image"],
+            qemu_args["cmdline"],
+            qemu_args["mem_size"],
+            qemu_args["num_cores"],
+            qemu_args["num_threads"],
+            self.connection_settings["port"],
+        )
+
+        if qemu_args["initrd_image"]:
+            if not os.path.exists(qemu_args["initrd_image"]):
+                raise FileNotFoundError(f'{qemu_args["initrd_image"]} does not exist!')
+
+            qemu_cmd += f' -initrd {qemu_args["initrd_image"]}'
+
+        if qemu_args["arch"] == machine():
+            if qemu_args["enable_kvm"]:
+                qemu_cmd += ' --enable-kvm'
+        else:
+            qemu_cmd += f' -machine virt -cpu {qemu_args["cpu_type"]}'
+
+        self.target = make_target(connect=False,
+                                  conn_cls=SshConnection,
+                                  connection_settings=self.connection_settings)
+
+        super().__init__(runner_cmd=qemu_cmd,
```

Comment:
All user input that makes it to qemu_cmd must be escaped using shlex.quote(). I know this is cumbersome, but the only other way to deal with that problem is to ask for the command as a list like `subprocess` does by default. `devlib` does not for various reasons (although we could easily add such an API, maybe we should ...)

Reply:
I thought this is addressed by `self.runner_process = get_subprocess(list(runner_cmd.split()))` in `TargetRunner` class?

Reply:
> get_subprocess(list(runner_cmd.split()))

This is broken. Either the `runner_cmd` is already a list and nothing needs to be split, or it's not and it needs to be split with `shlex.split()`. Just using `str.split()` will wreck any escaping that was applied. So considered the situation, we should be manipulating only lists here, and `qemu_cmd` should be a list as well. We _could_ use strings and then use `shlex.split()` but that's pointless and will just require extra work to create the string (shlex.quote on every single user input)

Reply:
Converted `qemu_cmd` to a list of string/parameters.